voxblox_ros/include/voxblox_ros/mesh_vis.h

```diff
@@ -147,31 +149,45 @@ inline void generateVoxbloxMeshMsg(const MeshLayer::Ptr& mesh_layer,
     mesh_block.index[1] = block_index.y();
     mesh_block.index[2] = block_index.z();
 
-    for (size_t i = 0u; i < mesh->vertices.size(); i += 3u) {
-      voxblox_msgs::Triangle triangle;
+    mesh_block.x.reserve(mesh->vertices.size());
+    mesh_block.y.reserve(mesh->vertices.size());
+    mesh_block.z.reserve(mesh->vertices.size());
 
-      for (size_t local_vert_idx = 0u; local_vert_idx < 3; ++local_vert_idx) {
-        const size_t global_vert_idx = local_vert_idx + i;
-        triangle.x[local_vert_idx] = mesh->vertices[global_vert_idx].x();
-        triangle.y[local_vert_idx] = mesh->vertices[global_vert_idx].y();
-        triangle.z[local_vert_idx] = mesh->vertices[global_vert_idx].z();
-
-        std_msgs::ColorRGBA color_msg =
-            getVertexColor(mesh, color_mode, global_vert_idx);
-
-        triangle.r[local_vert_idx] = static_cast<uint8_t>(255 * color_msg.r);
-        triangle.g[local_vert_idx] = static_cast<uint8_t>(255 * color_msg.g);
-        triangle.b[local_vert_idx] = static_cast<uint8_t>(255 * color_msg.b);
-        triangle.a[local_vert_idx] = 255;
+    // normal coloring is used by RViz plugin by default, so no need to send it
+    if (color_mode != kNormals) {
+      mesh_block.r.reserve(mesh->vertices.size());
+      mesh_block.g.reserve(mesh->vertices.size());
+      mesh_block.b.reserve(mesh->vertices.size());
+    }
+    for (size_t i = 0u; i < mesh->vertices.size(); ++i) {
+      // 0.5 is because mesh verticies can lie outside the block. We only
+      // strictly need one additional voxel of space (as opposed to one block
+      // given here), but communicating how many voxels there are per block
+      // adds some more coupling.
+      Point normalized_verticies =
```

Comment:
I will improve the commenting, but as I understand it the values will always be greater than zero, its just that they can be up to 1 voxel over the edge of the block in the positive direction.

Reply:
I think you're right, if I remember  correctly marching cubes associates the cross-block triangles always with the block that way. Still, it's an assumption and it can easily be broken which will lead to invalid uint16_t numbers. `CHECK_GE`?